### PR TITLE
Multiapp reset for multiple specification of times

### PR DESCRIPTION
**Pull request by abarun22**
This PR is associated with the issue: refs https://github.com/idaholab/moose/issues/22489

## Reason
The current way of resetting multi-apps does not provide options to perform the reset for multiple number of main app times. Hence the parameter "reset_time" has been converted to accept vector of 'double' values that will signify the range of times at which the multiapps specified in "reset_apps" will be reset

## Design
A simple conversion of "reset_time" from scalar to vector has been done. In addition the default action for the non-specification of "reset_apps" parameter was missing. If not specified the reset will be done for all associated apps.

## Impact
The function "MultiApp::preTransfer" is subject to minimal modifications that effect this new feature.

### CODE REVIEW

**Comment by GiudGiud** [comment]
Pull requests may not be made to master, only to next.

I ll push some changes to this

**Review by GiudGiud — DISMISSED** [review]
This will need a test. 

Please be more careful with the rebase next time. The lines you deleted with the getAppInfo (I re-added them) were there for a good reason.

**Comment by abarun22** [comment]
My goodness. I thought that was shown as an old code during the rebase and so i would have done that. Sorry about that

**Comment by GiudGiud** [comment]
Finally had time to fix this up and add a test. 
My bad i had introduced a few errors when I fixed up the rebasing commit. They are fixed. now.

**Comment by GiudGiud** [comment]
@cticenhour if you have a minute could you please review this?

File: framework/src/multiapps/MultiApp.C [hunk 1]
```diff
@@ -253,15 +254,20 @@ MultiApp::MultiApp(const InputParameters & parameters)
...
-      (_reset_apps.size() == 0 && _reset_time < std::numeric_limits<Real>::max()))
+  if ((_reset_apps.size() > 0 && _reset_times.size() == 0) ||
+      (_reset_apps.size() == 0 && _reset_times.size() > 0))
     mooseError("reset_time and reset_apps may only be specified together");
+
+  // Check that the reset times are sorted by the user
+  auto sorted_times = _reset_times;
+  std::sort(sorted_times.begin(), sorted_times.end());
+  if (_reset_times.size() && _reset_times != sorted_times)
+    paramError("reset_time", "List of reset times must be sorted in increasing order");
```
Comment (cticenhour) [inline framework/src/multiapps/MultiApp.C:270]: It doesn't appear as though you are hitting this error case in your test - please add another to hit this for coverage.

File: framework/src/multiapps/MultiApp.C [hunk 2]
```diff
@@ -516,22 +523,39 @@ MultiApp::fillPositions()
 void
 MultiApp::preTransfer(Real /*dt*/, Real target_time)
 {
-  // First, see if any Apps need to be Reset
-  if (!_reset_happened && target_time + 1e-14 >= _reset_time)
+  // Get a transient executioner to get a user-set tolerance
+  Real timestep_tol = 1e-13;
```
Comment (cticenhour) [inline framework/src/multiapps/MultiApp.C:527]: Is there a reason why this was changed from `1e-14`?

File: framework/src/multiapps/MultiApp.C [hunk 3]
```diff
@@ -516,22 +523,39 @@ MultiApp::fillPositions()
...
+  if (dynamic_cast<Transient *>(_fe_problem.getMooseApp().getExecutioner()))
+    timestep_tol =
+        dynamic_cast<Transient *>(_fe_problem.getMooseApp().getExecutioner())->timestepTol();
```
Comment (cticenhour) [inline framework/src/multiapps/MultiApp.C:530]: Probably will need to include `Transient.h` here to do this `dynamic_cast`. Seeing a failure here: https://civet.inl.gov/job/1280286/

**Review by cticenhour — CHANGES_REQUESTED** [review]
You've still got some modules failures (see CIVET results). Will give a pre-emptive review anyway since this is small and I noticed a few things.

File: framework/src/multiapps/MultiApp.C (hunk 2 above)
Comment (GiudGiud) [reply framework/src/multiapps/MultiApp.C:527]: default s changed for transient multiapps too. It s meant to be overriden anyway by what is set for the executioner

File: framework/src/multiapps/MultiApp.C (hunk 3 above)
Comment (GiudGiud) [reply framework/src/multiapps/MultiApp.C:530]: ah yes

**Comment by GiudGiud** [comment]
All green now! and review should be addressed

**Comment by cticenhour** [comment]
You have addressed my previous comment, but it looks like there is a possible related failure in Isopod (https://civet.inl.gov/job/1280434/). Though, this is the new Optimization Module right?

**Comment by GiudGiud** [comment]
This object is duplicated in the optimization module.
@lynnmunday can it be removed from isopod?

**Comment by GiudGiud** [comment]
Ah actually it's not duplicated, it just didnt get merged in the module.

Do you still need it?
I can patch it easily if needed

**Comment by lynnmunday** [comment]
@GiudGiud I don't know which object is duplicated?  I don't think we need it in isopod.

**Review by cticenhour — APPROVED** [review]
Comments addressed and fix for isopod noted in separate issue to be addressed during optimization module transfer.

**Comment by GiudGiud** [comment]
Thanks for the prompt review!

**Comment by GiudGiud** [comment]
also thanks for the contribution @abarun22 !

**Comment by abarun22** [comment]
Thanks Guillaume

**Pull request merged**